pkg/cloudprovider/providers/vsphere/vsphere.go

```
@@ -426,3 +438,243 @@ func (vs *VSphere) Routes() (cloudprovider.Routes, bool) {
+	vmDevices, err := vm.Device(ctx)
+	if err != nil {
+		return "", "", err
+	}
+
+	diskController, err := vmDevices.FindDiskController(vs.cfg.Disk.DiskController)
+	if err != nil {
+		if err.Error() == "no available SCSI controller" {
+			newSCSIController, err := vmDevices.CreateSCSIController(vs.cfg.Disk.SCSIControllerType)
+			if err != nil {
+				return "", "", err
+			}
+
+			configNewSCSIController := newSCSIController.(types.BaseVirtualSCSIController).GetVirtualSCSIController()
+			hotAndRemove := true
+			configNewSCSIController.HotAddRemove = &hotAndRemove
+			configNewSCSIController.SharedBus = types.VirtualSCSISharing(types.VirtualSCSISharingNoSharing)
+			err = vm.AddDevice(context.TODO(), newSCSIController)
+			if err != nil {
+				return "", "", err
+			}
+			vmDevices, err = vm.Device(ctx)
+			if err != nil {
+				return "", "", err
+			}
+			diskController, err = vmDevices.FindDiskController(vs.cfg.Disk.DiskController)
+			if err != nil {
+				return "", "", err
+			}
+		} else {
+			return "", "", err
+		}
+	}
+
+	disk := vmDevices.CreateDisk(diskController, ds.Reference(), vmDiskPath)
+	backing := disk.Backing.(*types.VirtualDiskFlatVer2BackingInfo)
+	backing.DiskMode = string(types.VirtualDiskModeIndependent_persistent)
+	disk = vmDevices.ChildDisk(disk)
+
+	err = vm.AddDevice(ctx, disk)
+	if err != nil {
+		return "", "", err
+	}
+
+	vmDevices, err = vm.Device(ctx)
+	if err != nil {
+		return "", "", err
+	}
+
+	devices := vmDevices.SelectByType(disk)
+	newDevice := devices[len(devices)-1]
+
+	diskUUID = getVirtualDiskUUID(newDevice)
+
+	return devices.Name(newDevice), diskUUID, nil
+}
+
+func getVirtualDiskUUID(newDevice types.BaseVirtualDevice) string {
+	vd := newDevice.GetVirtualDevice()
+
+	if b, ok := vd.Backing.(*types.VirtualDiskFlatVer2BackingInfo); ok {
+		uuidWithNoHypens := strings.Replace(b.Uuid, "-", "", -1)
+		return uuidWithNoHypens
+	}
+	glog.Warningf("Disk UUID not found! Mount will fail if there is no valid UUID")
```

if `mount` will fail later, would it be more appropriate to fail early by throwing caller an error?